### DISCUSSION TITLE
Looking for NeoGeo Pocket owners/testers

### DISCUSSION THREAD

**Discussion by PsyK0p4T** [sanni/cartreader#184] [Coding Corner]
Hi people. I'm currently working on improving NGP compatibility.
Thanks to splash5's original code, some 8M roms were already dumpable. Huge thanks to him.
I already managed to add 4M, more 8M, and 16M roms compatibility. Just before submitting everything, I'm missing 32M games which I don't have under my hands. So I'm looking for people owning :

- A sanni's cartreader
- A way to plug NGP cartridges to it (NGP-to-SNES adapter)
- Original cartridges of :
> Densha de Go! 2 on Neo Geo Pocket (NEOP0060)
> Metal Slug - 2nd Mission (NEOP0061)
> SNK vs. Capcom - The Match of the Millennium (NEOP0069)

If you want to help on this topic, send me a shout :)

**Comment by PsyK0p4T** [comment]
Just updated NGP.ino code.
The majority of 4, 8 and 16 Mbits chips should now be dumped fine. Tested on more than 30 retail games, and even a prototype.

I just don't have any 32M games at home to fully confirm if it works too. It would be cool to have some testing reports before a merge. Let me know if everything runs fine...

**Reply by majorpbx** [reply]
I have the board to dump those carts but alas no sockets can be found online and attempts at broken units for cheap to salvage the connector have gone in vain! Would you happen to know a source of supply? If I can get a socket and obtain one of those games I can help.

**Reply by PsyK0p4T** [reply]
Hey dude, thanks for your interest :)
I wasn't able to find NGP cartridge socket neither. Spare ones aren't sold anywhere it seems. Current options would be : 

to unsolder an original one from a console. Dead ones are cheaper but not so common. I took mine from a working but damaged and badly-looking unit.
Pros : easy to do. will fit perfectly on the adapter/pcb
Cons : could be pricey depending of the source console you'll get

to build your own one from a PCI socket. [link](https://hackaday.com/2012/06/20/machining-cartridge-connectors-from-pci-sockets/)
Pros : Really cheap solution. You can order many if you are afraid to fail on first one.
Cons : you'll have a bit of work to design it. And once the socket will be soldered on the adapter/pcb, the cartridges will plug horizontally instead of vertically

to steal one from various tools/flashing stuff designed for the NGP flashcards. Like Tengu adapters, FlashMasta etc.. but they could be even more rare and pricey online.

**Reply by majorpbx** [reply]
That's what my research has proven as well. I'm working on obtaining "SNK vs. Capcom - The Match of the Millennium (NEOP00690)" Japanese version at the moment. Once I get a cheap unit I'll salvage the connector. This might take a while so if anyone else can lend a hand feel free. I'll post here when I've secured everything but may take more time to get the items in-hand.

**Comment by uu36** [comment]
I am very grateful that you have supported 4M and 16M cartridges. I have Cartreader and NGP Adapter. I have "Densha de Go! 2 on Neo Geo Pocket (NEOP0060)" and
"Metal Slug - 2nd Mission (NEOP0061)". NEOP0060 is 32M, so I couldn't dump it with your latest ngp.ino. NEOP0061 is 16M, so I could dump it.

**Reply by PsyK0p4T** [reply]
Really all greetings must go to _splash5_  for all his initial code, dumping process etc. I just increased the compatibility and more chips support, but it wasn't the hardest part ;)

I still don't have any 32M games in hands. So at the moment it's a bit difficult for me to help on this. But I've checked the roms floating around, and these 3 games really seem to be 32M. NEOP0061 (Metal slug 2nd mission) looks like to include 2x16M Toshiba chips [according to this pic](http://ngpc.freeplaytech.com/wp-content/uploads/2011/02/Pic_of_3_cart_PCBs.jpg)
So maybe the cartridge response is returning the same result as for a 1x16M Toshiba chip, and the dumper read it as a 16M cart.

Do you have an email, discord, or something where I can contact you ? I would be interested by pictures of the PCBs. And I can make a debug version of NGP.ino for you to better identify and maybe support these games.

**Reply by uu36** [reply]
Thanks for the reply. Of course I am grateful to splash5 as well. I disassembled the cartridge and took some pictures of the pcb.
If you create a debug version of ngp.ino, please send it to dq9freak@gmail.com
![A3E2F891-E51F-45AA-A599-301C01408069](https://user-images.githubusercontent.com/87904238/126884240-6bb22f08-9542-4e67-8f68-82c22d5f21c8.jpeg)
![350EECB6-C34B-4359-A390-972921186B5C](https://user-images.githubusercontent.com/87904238/126884242-451181c9-f9a4-4fb1-af3a-2b6e988a6c60.jpeg)
I will test it.

**Reply by PsyK0p4T** [reply]
Thanks a lot for the fast answer and pics ! That's what I was expecting.
I got a debug version ready, I'm contacting you :)

**Reply by majorpbx** [reply]
I finally have the game and a completed NGP adapter. Still need help testing? Though my setup keeps saying "Rom Size Error" and I'm pretty sure the one game I'm testing is 16M or less as it has 1 chip.

**Reply by PsyK0p4T** [reply]
WS adapter ? or NGP ? ^^
- First, try to clean the pins of your cartridge with alcohol and q-tip. Then retry.
- check if your cartridge reader is up to date, and your solders correct on your adapter.
- Post here the title of the games you're trying to dump, a picture of the pcb would help too.

**Reply by majorpbx** [reply]
Updated my comment as that was a slip on my part! It doesn't matter which cart I try to dump, once I go into the option for NGP it automatically says "Rom Size Error". I have cleaned the contacts and the connector's contacts and verified the solder joints.

Game is  King of Fighters R-2 (J)

![S20210810_0001](https://user-images.githubusercontent.com/18235905/128912063-721f9d82-c803-4942-8d5a-67fb6c277ffb.jpg)

**Reply by PsyK0p4T** [reply]
King of Fighters R-2 (NEOP0023) was released worldwide with the exact same content. I got an EUR cartridge here, you got a JPN one, but both seem to contain a 16 mbit Toshiba chip. This one should be handled fine by the NGP dumper currently on this github. At least I can dump mine without problem with it. Some checks you can do :
- Make sure you downloaded the latest "NGP.ino" file from here, and flashed it to your reader
- Make sure the switches are all pushed to the left (3V, and no EEP/CLK0/CLK1)
- Make sure your NGP adapter is perfectly aligned with the SNES slot pins too, and pins clean
- Try to brush once more the cartridge pins with alcohol and q-tip

If it continues to fail, I'll send you a debug version. It could be an unknown chip, but I don't think so.
Have you managed to dump other carts ? Or KOF-R2 is the only one you have at the moment ?

**Reply by majorpbx** [reply]
Updating the NGP.ino was indeed the solution! Thank you and if you need me to test my copy of SNK vs. Capcom - The Match of the Millennium (NEOP0069) let me know.

**Reply by PsyK0p4T** [reply]
@uu36 @majorpbx after the latest commit I think you can now try again your copies of Metal Slug 2nd Mission, Densha de Go, or SNK vs Capcom Match of the Millennium. They should be good.

**Comment by PsyK0p4T** [comment]
Ok so, I finally received a copy of Metal Slug 2nd Mission. As already discussed here, this one is a 32 mbits game, designed with dual 16 mbits flash chips inside. Due to this, chip ID commands return the same infos as the single-chip 16 mbits games. 
So in the latest commit I modified the code a bit, to deal with the 3 concerned titles correctly among other things. 32 mbits games should now be dumped fine.

I also tried to use another method to query the flash chip capacity directly by using extended CFI commands. The code would be something like this :

dataOut();
writeByte_NGP(0x00, 0xf0); //reset
writeByte_NGP(0x55, 0x98); //enter CFI mode
dataIn();                
tmp = readByte_NGP(0x27);

here "tmp" would normally take values 0x13 (for 512kb / 4 Mbits chips), 0x14 (1MB / 8Mbits) or 0x15 (2MB / 16 Mbits). It actually worked, but only for one game (Faselei! NEOP0051). All other cartridges I tested, more than 20 different ones, returned unexpected values. So I finally dropped this option for now. If someone is interested to investigate further on this, please do ;)

**Reply by majorpbx** [reply]
Awesome! I'll give it a try when I can and report back.
